### PR TITLE
DynamicPackageOutputManager and its implementation

### PR DESCRIPTION
**Pull request by ratcashdev**
As a preparation, and with the assumption that https://github.com/real-logic/simple-binary-encoding/pull/904 will receive no major changes, I'd like to propose to merge this PR for the upcoming version of agrona.

### CODE REVIEW

File: agrona/src/main/java/org/agrona/generation/MultiPackageOutputManager.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.agrona.generation;
+
+/**
+ * Extended version of the OutputManager allowing to set the java package.
+ */
+public interface MultiPackageOutputManager extends OutputManager
```
[the remark this answers is not in the document]
Comment (ratcashdev) [reply agrona/src/main/java/org/agrona/generation/MultiPackageOutputManager.java:21]: packages are used to group, organize or distinguish. In this sense the following comes to my mind:
`(Code)OrganizingOutputManager`,  `DynamicPackageOutputManager`,  `GroupingOutputManager`, `PackageControlledOutputManager`,  `SwitchablePackageOutputManager`, `SelectablePackageOutputManager`

File: agrona/src/main/java/org/agrona/generation/MultiPackageOutputManager.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /**
+     * Sets the current package name.
```
[the remark this answers is not in the document]
Comment (ratcashdev) [reply agrona/src/main/java/org/agrona/generation/MultiPackageOutputManager.java]: javadoc added.

File: agrona/src/main/java/org/agrona/generation/StringWriterOutputManager.java [hunk 3]
```diff
@@ -36,7 +39,15 @@ public Writer createOutput(final String name)
         final StringWriter stringWriter = new StringWriter();
         sourceFileByName.put(packageName + "." + name, stringWriter);
 
-        return stringWriter;
+        return new FilterWriter(stringWriter)
+        {
+            @Override
```
Comment (mjpt777) [inline agrona/src/main/java/org/agrona/generation/StringWriterOutputManager.java:44]: We do not use the @Override annotation.

File: agrona/src/main/java/org/agrona/generation/PackageGovernedOutputManager.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.agrona.generation;
+
+/**
+ * Extended version of the OutputManager allowing to specify packages for selected outputs.
+ */
+public interface PackageGovernedOutputManager extends OutputManager
```
Comment (mjpt777) [inline agrona/src/main/java/org/agrona/generation/PackageGovernedOutputManager.java:21]: Sorry but this name jars.

File: agrona/src/main/java/org/agrona/generation/MultiPackageOutputManager.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjpt777) [reply agrona/src/main/java/org/agrona/generation/MultiPackageOutputManager.java]: `DynamicPackageOutputManger` is probably one of the better.

[the remark this answers is not in the document]
Comment (ratcashdev) [reply agrona/src/main/java/org/agrona/generation/MultiPackageOutputManager.java]: OK, renamed.

File: agrona/src/main/java/org/agrona/generation/StringWriterOutputManager.java (hunk 3 above)
Comment (ratcashdev) [reply agrona/src/main/java/org/agrona/generation/StringWriterOutputManager.java]: yes, keep forgetting, apologies.

**Comment by ratcashdev** [comment]
@mjpt777 is there a way to reference the merged (snapshot) version of agrona from SBE's PR? Or should I just wait a couple of days till a release is published?

**Comment by mjpt777** [comment]
> @mjpt777 is there a way to reference the merged (snapshot) version of agrona from SBE's PR? Or should I just wait a couple of days till a release is published?

This can only be done locally.

**Pull request merged**